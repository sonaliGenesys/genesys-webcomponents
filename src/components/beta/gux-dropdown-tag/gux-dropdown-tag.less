@import (reference) '../../../style/color.less';
@import (reference) '../../../style/mixins.less';
@import (reference) '../../../style/shadows.less';
@import (reference) '../../../style/spacing.less';
@import (reference) '../../../style/typography.less';

:host {
  color: @gux-black-50;
}

.gux-dropdown-container {
  position: relative;
}

.gux-error {
  &.gux-target-container-collapsed .gux-field-button,
  &.gux-target-container-expanded {
    border-color: @gux-alert-red-60;
  }
}

.gux-field,
.gux-target-container-expanded {
  all: unset;
  box-sizing: border-box;
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  align-content: stretch;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 32px;
  padding: 6px 0 6px 6px;
  font-family: inherit;
  font-size: 12px;
  line-height: 1.6667;
  cursor: pointer;
  background-color: @gux-grey-90;
}

.gux-field {
  &.gux-input-field {
    height: 30px;
    cursor: text;
  }

  .gux-field-content {
    position: relative;
    display: flex;
    flex: 1 1 0;
    flex-direction: row;
    flex-wrap: nowrap;
    align-content: stretch;
    align-items: flex-start;
    justify-content: flex-start;
    min-width: 0;
    height: 100%;

    .gux-filter,
    .gux-selected-option,
    .gux-placeholder {
      flex: 1 1 auto;
      align-self: auto;
      order: 0;
      padding-left: 6px;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;

      .gux-sr-only {
        .gux-sr-only-clip();
      }
    }

    .gux-filter {
      position: relative;
      height: 100%;
      padding-left: 0;

      .gux-filter-input {
        all: unset;
        width: 100%;
        caret-color: @gux-black-50;
      }

      .gux-filter-display {
        white-space: pre;

        .gux-filter-text {
          color: @gux-black-50;
        }

        .gux-filter-suggestion {
          color: @gux-black-90;
        }
      }

      .gux-filter-input,
      .gux-filter-display {
        position: absolute;
      }
    }

    .gux-placeholder {
      color: @gux-black-90;
    }
  }

  .gux-expand-icon {
    flex: 0 0 auto;
    align-self: auto;
    order: 0;
    width: 16px;
    height: 16px;
    padding: 0 @gux-spacing-xs;
    color: @gux-black-90;
  }
}

.gux-target-container-expanded {
  .form-field-border();

  &:focus-within {
    .focus-ring();
  }

  .gux-filter-input {
    background-color: inherit;
    border: none;

    &:focus {
      border: none;
      outline: none;
    }
  }

  .gux-field-button {
    width: auto;
    height: 30px;
    padding: 0 0 0 @gux-spacing-xs;
    margin: 0;
    background: inherit;
    border: none;
    outline: none;
    box-shadow: none;

    &:focus {
      outline: none;
    }
  }
}

.gux-target-container-collapsed .gux-field-button {
  .form-field-border();

  &:focus {
    .focus-ring();
  }
}

.gux-listbox-container {
  box-sizing: border-box;
  margin: 0;
  color: @gux-black-50;
}

.gux-error-message-container {
  display: none;
  margin: 4px 0;

  &.gux-show {
    display: flex;
  }
}
