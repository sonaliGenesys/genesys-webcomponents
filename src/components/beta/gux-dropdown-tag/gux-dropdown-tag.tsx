import {
  Component,
  Element,
  Event,
  EventEmitter,
  forceUpdate,
  h,
  JSX,
  Listen,
  Method,
  Prop,
  State,
  Watch
} from '@stencil/core';

import { OnClickOutside } from '../../../utils/decorator/on-click-outside';
import { buildI18nForComponent, GetI18nValue } from '../../../i18n';
import simulateNativeEvent from '../../../utils/dom/simulate-native-event';
import { afterNextRender } from '../../../utils/dom/after-next-render';
import { onInputDisabledStateChange } from '../../../utils/dom/on-input-disabled-state-change';
import { trackComponent } from '../../../usage-tracking';

import translationResources from './i18n/en.json';

import { getSearchOption } from '../../stable/gux-listbox/gux-listbox.service';
import { GuxFilterTypes } from '../../stable/gux-dropdown/gux-dropdown.types';
/**
 * @slot - for a gux-listbox-multi containing gux-option-multi children
 */
@Component({
  styleUrl: 'gux-dropdown-tag.less',
  tag: 'gux-dropdown-tag-beta',
  shadow: true
})
export class GuxDropdownTag {
  private i18n: GetI18nValue;
  private fieldButtonElement: HTMLElement;
  private textInputElement: HTMLInputElement;
  private listboxElement: HTMLGuxListboxMultiElement;

  @Element()
  private root: HTMLGuxDropdownMultiBetaElement;

  @Prop({ mutable: true })
  value: string;

  @Prop()
  disabled: boolean = false;

  @Prop()
  required: boolean = false;

  @Prop()
  loading: boolean = false;

  @Prop()
  placeholder: string;

  /**
   * deprecated will be removed in v4 (COMUI-1369). Use filterType instead
   */
  @Prop()
  filterable: boolean = false;

  /**
   * Override default filtering behavior
   */
  @Prop()
  filterType: GuxFilterTypes = 'none';

  @Prop()
  hasError: boolean = false;

  @State()
  hasCreate: boolean = false;

  @State()
  private expanded: boolean = false;

  @State()
  private textInput: string = '';

  /**
   * This event is emitted to request creating a new option
   */
  @Event()
  guxcreateoption: EventEmitter;

  /**
   * This event will run when the dropdown-multi transitions to an expanded state.
   */
  @Event()
  guxexpanded: EventEmitter<void>;

  /**
   * This event will run when the dropdown-multi transitions to a collapsed state.
   */
  @Event()
  guxcollapsed: EventEmitter<void>;

  @Event()
  private guxfilter: EventEmitter<string>;

  /**
   * Listens for expanded event emitted by gux-popup.
   */
  @Listen('internalexpanded')
  onInternalExpanded(event: CustomEvent): void {
    event.stopPropagation();
    this.guxexpanded.emit();
  }

  /**
   * Listens for collapsed event emitted by gux-popup.
   */
  @Listen('internalcollapsed')
  onInternalCollapsed(event: CustomEvent): void {
    event.stopPropagation();
    this.guxcollapsed.emit();
  }

  @Watch('expanded')
  focusSelectedItemAfterRender(expanded: boolean) {
    if (expanded && this.listboxElement) {
      afterNextRender(() => {
        if (this.hasTextInput() && this.textInputElement) {
          this.textInputElement.focus();
        } else {
          this.listboxElement.focus();
        }
      });
    }

    if (!expanded) {
      this.textInput = '';
    }
  }

  @Watch('value')
  validateValue(newValue: string) {
    if (newValue === undefined) {
      if (this.listboxElement) {
        this.listboxElement.value = newValue;
      }
      return;
    }

    const selectedListboxOptionElement = this.getOptionElementByValue(newValue);

    if (selectedListboxOptionElement) {
      this.listboxElement.value = newValue;
      return;
    }

    this.value = undefined;
  }

  @Watch('textInput')
  handleFilter(filter: string) {
    this.guxfilter.emit(filter);
  }

  /**
   * Returns an array of the selected values
   */
  @Method()
  getSelectedValues(): Promise<string[]> {
    return Promise.resolve(this.value?.split(',') || []);
  }

  @Listen('keydown')
  onKeydown(event: KeyboardEvent): void {
    switch (event.key) {
      case 'Escape':
        if (this.hasTextInput()) {
          if (document.activeElement === this.listboxElement) {
            return this.textInputElement.focus();
          }
        }
        this.collapseListbox('focusFieldButton');
        return;
      case 'Tab':
        if (this.shiftTabFromFilterListbox(event)) {
          event.preventDefault();
          return this.textInputElement.focus();
        } else if (this.shiftTabFromExpandedFilterInput(event)) {
          event.preventDefault();
          return this.collapseListbox('focusFieldButton');
        } else {
          this.collapseListbox('noFocusChange');
        }
        return;
      case 'ArrowDown':
        if (this.activeElementNotListbox()) {
          event.preventDefault();
          this.expanded = true;
        }
        return;
      case 'Enter':
        if (this.canCreateNewOption() && this.isActiveElement()) {
          this.emitCreateOption();
        }
    }
  }

  /**
   * force update when slotted gux-listbox-multi listbox options change
   */
  @Listen('internallistboxoptionsupdated')
  onInternallistboxoptionsupdated(event: CustomEvent): void {
    event.stopPropagation();
    forceUpdate(this.root);
  }

  /**
   * clear selected options when gux-dropdown-tag-value emits event
   */
  @Listen('tagCloseClicked')
  onClearselected(event: CustomEvent): void {
    event.stopPropagation();

    const clearValue: string = event.detail;
    this.value = this.removeValue(this.value, clearValue);

    this.updateValue(this.value);
    if (this.listboxElement) {
      this.listboxElement.value = this.value;
    }
    this.fieldButtonElement.focus();

    this.expanded = this.value ? true : false;
  }

  removeValue(value: string, clearValue: string): string {
    const arrayElements = value.split(',');
    const index = arrayElements.indexOf(clearValue);

    if (index !== -1) arrayElements.splice(index, 1);

    return arrayElements.toString();
  }

  /**
   * emit guxcreateoption event when gux-create-option emits create event
   */
  @Listen('internalcreatenewoption')
  onCreatenewoption(event: CustomEvent): void {
    event.stopPropagation();
    this.emitCreateOption();
  }

  @Listen('blur')
  onBlur(event: FocusEvent): void {
    this.stopPropagationOfInternalFocusEvents(event);
  }

  @Listen('focus')
  onFocus(event: FocusEvent): void {
    this.stopPropagationOfInternalFocusEvents(event);
  }

  @Listen('focusout')
  onFocusout(event: FocusEvent): void {
    this.stopPropagationOfInternalFocusEvents(event);
  }

  @Listen('focusin')
  onFocusin(event: FocusEvent): void {
    this.stopPropagationOfInternalFocusEvents(event);
  }

  @OnClickOutside({ triggerEvents: 'mousedown' })
  onClickOutside() {
    this.collapseListbox('noFocusChange');
  }

  connectedCallback(): void {
    this.listboxElement = this.root.querySelector('gux-listbox-multi');
    this.hasCreate = !!this.root.querySelector('gux-create-option');
    this.validateValue(this.value);
  }

  async componentWillLoad(): Promise<void> {
    trackComponent(this.root);
    this.i18n = await buildI18nForComponent(this.root, translationResources);

    onInputDisabledStateChange(this.root, () => {
      forceUpdate(this.root);
    });
  }

  componentDidLoad(): void {
    this.listboxElement.addEventListener('input', (event: InputEvent) => {
      event.stopPropagation();

      this.updateValue((event.target as HTMLGuxListboxMultiElement).value);
    });
    this.listboxElement.addEventListener('change', (event: InputEvent) => {
      event.stopPropagation();
    });
  }

  componentWillRender(): void {
    this.validateValue(this.value);
    this.listboxElement.loading = this.loading;
    this.listboxElement.filterType = this.filterType;
    this.listboxElement.textInput = this.textInput;
  }

  private hasTextInput(): boolean {
    return this.isFilterable() || this.hasCreate;
  }

  private isFilterable(): boolean {
    return (
      this.filterable ||
      this.filterType === 'custom' ||
      this.filterType === 'starts-with'
    );
  }

  private stopPropagationOfInternalFocusEvents(event: FocusEvent): void {
    if (this.root.contains(event.relatedTarget as Node)) {
      return event.stopImmediatePropagation();
    }
  }

  private getOptionElementByValue(value: string): HTMLGuxOptionElement[] {
    const listboxOptionElements = Array.from(
      this.root.querySelectorAll('gux-option-multi')
    );

    const values = value ? value.split(',') : undefined;
    if (values) {
      return listboxOptionElements.filter(element =>
        values.includes(element.value)
      );
    }
    return;
  }

  private fieldButtonClick(): void {
    this.expanded = !this.expanded;
  }

  private fieldButtonInputClick(): void {
    if (!this.expanded) {
      this.expanded = !this.expanded;
    }
  }

  private filterInput(event: InputEvent): void {
    event.stopPropagation();
    this.textInput = this.textInputElement.value;
  }

  private shiftTabFromExpandedFilterInput(event: KeyboardEvent): boolean {
    return (
      event.shiftKey &&
      this.hasTextInput() &&
      this.expanded &&
      !(document.activeElement === this.listboxElement)
    );
  }

  private shiftTabFromFilterListbox(event: KeyboardEvent): boolean {
    return (
      event.shiftKey &&
      this.hasTextInput() &&
      document.activeElement === this.listboxElement
    );
  }

  private emitCreateOption(): void {
    this.guxcreateoption.emit(this.textInput);
    this.textInput = '';
    this.textInputElement.value = '';
  }

  /**
   * check if able to create new option from text input value
   */
  private canCreateNewOption(): boolean {
    return (
      this.hasCreate && this.textInput && !this.listboxElement.hasExactMatch
    );
  }

  private isActiveElement(): boolean {
    return document.activeElement === this.root;
  }

  private activeElementNotListbox(): boolean {
    return document.activeElement !== this.listboxElement;
  }

  private filterKeydown(event: KeyboardEvent): void {
    switch (event.key) {
      case 'ArrowDown':
        event.stopImmediatePropagation();
        this.listboxElement.focus();
        return;
    }
  }

  private filterKeyup(event: KeyboardEvent): void {
    switch (event.key) {
      case ' ':
        event.preventDefault();
        return;
    }
  }

  private collapseListbox(
    focusChange: 'noFocusChange' | 'focusFieldButton'
  ): void {
    if (this.expanded) {
      this.expanded = false;
    }

    if (focusChange === 'focusFieldButton') {
      this.fieldButtonElement.focus();
    }
  }

  private updateValue(newValue: string): void {
    if (this.value !== newValue) {
      this.value = newValue;
      simulateNativeEvent(this.root, 'input');
      simulateNativeEvent(this.root, 'change');
    }
  }

  private getTypeaheadText(textInput: string): string {
    const textInputLength = textInput.length;
    if (textInputLength > 0 && !this.loading) {
      const option = getSearchOption(this.listboxElement, textInput);
      if (option && this.filterType !== 'custom') {
        const optionSlotTextContent = option.querySelector(
          '[gux-slot-container]'
        )?.textContent;
        return optionSlotTextContent?.substring(textInputLength);
      }

      return '';
    }
  }

  private renderTargetDisplay(): JSX.Element {
    return (
      <div class="gux-placeholder">
        {!(this.expanded && this.hasTextInput()) &&
          (this.placeholder || this.i18n('noSelection'))}
      </div>
    ) as JSX.Element;
  }

  private getInputAriaLabel(): string {
    return this.canCreateNewOption() && this.isActiveElement()
      ? this.i18n('pressEnterToCreate', { textInputValue: this.textInput })
      : this.i18n('textInputResults');
  }

  private renderTag(): JSX.Element[] {
    const selectedListboxOptionElement = this.getOptionElementByValue(
      this.value
    );

    if (selectedListboxOptionElement?.length) {
      const optionsValue = [];
      for (const product of selectedListboxOptionElement) {
        optionsValue.push(
          <gux-dropdown-tag-value
            disabled={this.disabled}
            option-selected={product.textContent}
            value={product.value}
          ></gux-dropdown-tag-value>
        );
      }

      return optionsValue as JSX.Element[];
    }
  }

  private renderFilterInputFieldText(): JSX.Element {
    if (this.expanded && this.hasTextInput()) {
      const filterSuggestion = this.getTypeaheadText(this.textInput);
      return (
        <div class="input-container">
          <div class="gux-filter-display">
            <span class="gux-filter-text">
              {filterSuggestion !== '' ? this.textInput : ''}
            </span>
            <span class="gux-filter-suggestion">{filterSuggestion}</span>
          </div>
          <div class="input-and-dropdown-button">
            <input
              onClick={this.fieldButtonInputClick.bind(this)}
              // placeholder={!this.value && (this.placeholder || this.i18n('noSelection'))}
              class="gux-filter-input"
              type="text"
              aria-label={this.getInputAriaLabel()}
              ref={el => (this.textInputElement = el)}
              onInput={this.filterInput.bind(this)}
              onKeyDown={this.filterKeydown.bind(this)}
              onKeyUp={this.filterKeyup.bind(this)}
              autocomplete="off"
              autocorrect="off"
              autocapitalize="off"
              spellcheck="false"
              style={{ width: this.getInputWidth() }}
            ></input>
          </div>
        </div>
      ) as JSX.Element;
    }
  }

  /**
   * Returns a width for filter text input element
   */
  getInputWidth(): string {
    if (this.textInputElement)
      return ((this.textInputElement.value.length + 1) * 7).toString();
    return '20';
  }

  private renderPopup(): JSX.Element {
    return (
      <div slot="popup" class="gux-listbox-container">
        <slot />
      </div>
    ) as JSX.Element;
  }

  private renderTarget(): JSX.Element {
    return (
      <div
        class={{
          'gux-target-container': true,
          'gux-error': this.hasError
        }}
        slot="target"
      >
        <button
          type="button"
          class="gux-field gux-field-button"
          disabled={this.disabled}
          onClick={this.fieldButtonClick.bind(this)}
          ref={el => (this.fieldButtonElement = el)}
          aria-haspopup="listbox"
          aria-expanded={this.expanded.toString()}
        >
          {this.renderTargetContent()}
          {this.renderRadialLoading()}
          <gux-icon
            class={{
              'gux-expand-icon': true
            }}
            screenreader-text={this.i18n('dropdown')}
            iconName="chevron-small-down"
          ></gux-icon>
        </button>
      </div>
    ) as JSX.Element;
  }

  private renderTargetContent(): JSX.Element {
    return (
      <div class="gux-field-content">
        {this.value ? this.renderTag() : this.renderTargetDisplay()}
        {this.renderFilterInputFieldText()}
      </div>
    ) as JSX.Element;
  }

  private renderRadialLoading(): JSX.Element {
    if (this.loading && !this.expanded) {
      return (
        <gux-radial-loading context="input"></gux-radial-loading>
      ) as JSX.Element;
    }
  }

  render(): JSX.Element {
    return [
      <div class="gux-dropdown-container">
        <gux-popup
          expanded={this.expanded && (!this.loading || this.isFilterable())}
          disabled={this.disabled}
        >
          {this.renderTarget()}
          {this.renderPopup()}
        </gux-popup>
      </div>
    ] as JSX.Element;
  }
}
