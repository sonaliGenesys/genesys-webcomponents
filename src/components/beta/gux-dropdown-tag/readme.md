# gux-dropdown-multi

<!-- Auto Generated Below -->


## Properties

| Property      | Attribute     | Description                                                           | Type                                  | Default     |
| ------------- | ------------- | --------------------------------------------------------------------- | ------------------------------------- | ----------- |
| `disabled`    | `disabled`    |                                                                       | `boolean`                             | `false`     |
| `filterType`  | `filter-type` | Override default filtering behavior                                   | `"custom" \| "none" \| "starts-with"` | `'none'`    |
| `filterable`  | `filterable`  | deprecated will be removed in v4 (COMUI-1369). Use filterType instead | `boolean`                             | `false`     |
| `hasError`    | `has-error`   |                                                                       | `boolean`                             | `false`     |
| `loading`     | `loading`     |                                                                       | `boolean`                             | `false`     |
| `placeholder` | `placeholder` |                                                                       | `string`                              | `undefined` |
| `required`    | `required`    |                                                                       | `boolean`                             | `false`     |
| `value`       | `value`       |                                                                       | `string`                              | `undefined` |


## Events

| Event             | Description                                                                   | Type                  |
| ----------------- | ----------------------------------------------------------------------------- | --------------------- |
| `guxcollapsed`    | This event will run when the dropdown-multi transitions to a collapsed state. | `CustomEvent<void>`   |
| `guxcreateoption` | This event is emitted to request creating a new option                        | `CustomEvent<any>`    |
| `guxexpanded`     | This event will run when the dropdown-multi transitions to an expanded state. | `CustomEvent<void>`   |
| `guxfilter`       |                                                                               | `CustomEvent<string>` |


## Methods

### `getSelectedValues() => Promise<string[]>`

Returns an array of the selected values

#### Returns

Type: `Promise<string[]>`




## Slots

| Slot | Description                                                  |
| ---- | ------------------------------------------------------------ |
|      | for a gux-listbox-multi containing gux-option-multi children |


## Dependencies

### Depends on

- [gux-dropdown-tag-value](gux-dropdown-tag-value)
- [gux-icon](../../stable/gux-icon)
- [gux-radial-loading](../../stable/gux-radial-loading)
- [gux-popup](../../stable/gux-popup)

### Graph
```mermaid
graph TD;
  gux-dropdown-tag-beta --> gux-dropdown-tag-value
  gux-dropdown-tag-beta --> gux-icon
  gux-dropdown-tag-beta --> gux-radial-loading
  gux-dropdown-tag-beta --> gux-popup
  gux-dropdown-tag-value --> gux-icon
  style gux-dropdown-tag-beta fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
