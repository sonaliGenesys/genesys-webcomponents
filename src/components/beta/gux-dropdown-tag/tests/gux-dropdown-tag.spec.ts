import { newSpecPage } from '@stencil/core/testing';
import MutationObserver from 'mutation-observer';

import { GuxDropdownTag } from '../gux-dropdown-tag';
import { GuxListboxMulti } from '../../../stable/gux-listbox-multi/gux-listbox-multi';
import { GuxOptionMulti } from '../../../stable/gux-listbox-multi/gux-option-multi/gux-option-multi';

const components = [GuxDropdownTag, GuxListboxMulti, GuxOptionMulti];
const html = `
<gux-dropdown-tag-beta>
  <gux-listbox-multi aria-label="Animals">
    <gux-option-multi value="a" disabled>Ant</gux-option-multi>
    <gux-option-multi value="b">Bat</gux-option-multi>
    <gux-option-multi value="c">Cat</gux-option-multi>
  </gux-listbox-multi>
</gux-dropdown-tag-beta>
`;
const language = 'en';

describe('gux-dropdown', () => {
  beforeEach(() => {
    (
      global as NodeJS.Global & {
        MutationObserver: any;
      }
    ).MutationObserver = MutationObserver;
  });

  describe('#render', () => {
    it(`should render as expected`, async () => {
      const page = await newSpecPage({ components, html, language });

      expect(page.rootInstance).toBeInstanceOf(GuxDropdownTag);
      expect(page.root).toMatchSnapshot();
    });
  });
});
