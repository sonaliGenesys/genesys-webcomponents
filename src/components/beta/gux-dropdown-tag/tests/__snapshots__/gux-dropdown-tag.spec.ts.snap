// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-dropdown #render should render as expected 1`] = `
<gux-dropdown-tag>
  <mock:shadow-root>
    <div class="gux-dropdown-container">
      <gux-popup>
        <div class="gux-target-container gux-target-container-collapsed" slot="target">
          <button aria-expanded="false" aria-haspopup="listbox" class="gux-field gux-field-button" type="button">
            <div class="gux-field-content">
              <div class="gux-placeholder">
                Select...
              </div>
            </div>
            <gux-icon class="gux-expand-icon" iconname="chevron-small-down" screenreader-text="Dropdown"></gux-icon>
          </button>
        </div>
        <div class="gux-listbox-container" slot="popup">
          <slot></slot>
        </div>
      </gux-popup>
    </div>
  </mock:shadow-root>
  <gux-listbox-multi aria-label="Animals" aria-multiselectable="true" role="listbox" tabindex="0">
    <mock:shadow-root>
      <slot></slot>
      <slot name="create"></slot>
    </mock:shadow-root>
    <gux-option-multi aria-disabled="true" aria-selected="false" class="gux-disabled" disabled="" id="gux-option-multi-i" role="option" value="a">
      <!---->
      <div class="gux-option" gux-slot-container="">
        Ant
      </div>
    </gux-option-multi>
    <gux-option-multi aria-disabled="false" aria-selected="false" id="gux-option-multi-i" role="option" value="b">
      <!---->
      <div class="gux-option" gux-slot-container="">
        Bat
      </div>
    </gux-option-multi>
    <gux-option-multi aria-disabled="false" aria-selected="false" id="gux-option-multi-i" role="option" value="c">
      <!---->
      <div class="gux-option" gux-slot-container="">
        Cat
      </div>
    </gux-option-multi>
  </gux-listbox-multi>
</gux-dropdown-tag>
`;
